### PR TITLE
DAOS-7027 sched: Detecting non-yield ULTs

### PR DESCRIPTION
**Pull request by NiuYawei**
Add mechanism to detect ULTs who doesn't yield voluntarily or blocks
xstream unexpectedly (for instance, by system call). Some hack code
is introduced for this purpose, they should be replaced by Argobots
APIs if there are proper APIs exported in later version.

This patch also shrunk the SCHED_AGE_NET_MAX and SCHED_AGE_NVME_MAX
to 32 and 64 respectively.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

**Comment by NiuYawei** [comment]
There isn't an easy way to get function address from ABT_unit so far, I raised ticket for Argobots: https://github.com/pmodels/argobots/issues/314

Before the ABT API is ready, I used some dirty hack to get the function address. It worked in my test environment, so I think you may apply the patch for debugging if you want.

CI NLT test failed on starting server (segfault), I guess it could because the hack doesn't work with the binaries used by NLT, need more investigation on this.

**Comment by NiuYawei** [comment]
oops, my hack code is based on ABT 1.0, will update the patch.

**Review by wangdi1 — DISMISSED** [review]

File: src/engine/sched.c [hunk 1]
```diff
@@ -1544,6 +1548,89 @@ sched_start_cycle(struct sched_data *data, ABT_pool *pools)
...
+static void
+sched_watchdog_prep(struct dss_xstream *dx, ABT_unit unit,
+		    struct sched_unit *su)
+{
+	struct ABTI_thread_dup	*td = (struct ABTI_thread_dup *)unit;
+
+	if (!watchdog_enabled(dx))
+		return;
+
+	su->su_start = daos_getntime_coarse() / NSEC_PER_MSEC;
```
Comment (liw) [inline src/engine/sched.c:1593]: I recall you added daos_getmtime_coarse recently.

**Review by liw — APPROVED** [review]
Thanks, Niu.

File: src/engine/sched.c (hunk 1 above)
Comment (NiuYawei) [reply src/engine/sched.c:1593]: yeah, but that patch hasn't been landed yet.

**Comment by jolivier23** [comment]
I took the liberty of merging latest master to get the PR provisioning fix because you are pretty far back in the queue anyway.

**Comment by NiuYawei** [comment]
RPM build failed, seems RPM build update for new argobots isn't done yet? @brianjmurrell

**Comment by NiuYawei** [comment]
Needs to refresh patch once two RPM build PRs landed (tracked in DAOS-7024.)

**Review by liw — DISMISSED** [review]

**Review by wangdi1 — DISMISSED** [review]

**Comment by NiuYawei** [comment]
@daos-stack/daos-gatekeeper , all tests passed except that NLT result is "unstable", that because 5 new error messages generated by this patch, the error messages are expected, they are printed because some RPC handle ULT took unexpected long time (probably they are pool creating RPC).

File: src/engine/sched.c [hunk 2]
```diff
@@ -1551,6 +1555,92 @@ sched_start_cycle(struct sched_data *data, ABT_pool *pools)
...
+	D_ASSERT(cur >= su->su_start);
+	elapsed = cur - su->su_start;
+
+	if (elapsed <= sched_unit_runtime_max)
+		return;
+
+	/* Throttle printing a bit */
+	D_ASSERT(cur >= info->si_stats.ss_watchdog_ts);
+	if (info->si_stats.ss_last_unit == su->su_func_addr &&
+	    cur - info->si_stats.ss_watchdog_ts <= 2000)
```
Comment (jolivier23) [inline src/engine/sched.c:1628]: minor, I really prefer parentheses for readability.

**Review by jolivier23 — DISMISSED** [review]

**Comment by NiuYawei** [comment]
> @NiuYawei Are the errors consistent? Do the same number happen every time? If not, we may have an issue. Otherwise, @sylviachanoiyee or @johannlombardi can force land it.

Yes, my observation is that pool create ULT will usually take more than 100 ms, and last CI run NLT also reported the same 5 new error messages.

**Comment by ashleypittman** [comment]
> > @NiuYawei Are the errors consistent? Do the same number happen every time? If not, we may have an issue. Otherwise, @sylviachanoiyee or @johannlombardi can force land it.
> 
> Yes, my observation is that pool create ULT will usually take more than 100 ms, and last CI run NLT also reported the same 5 new error messages.

It may be that we need to mask the time out from the error message, as we do with pointers currently but if the errors are consistent then it should be OK.

I assume this is the first step to identifying and resolving long-running threads?  If so might it be more useful to dump the full stack in this case?

**Comment by NiuYawei** [comment]
> > > @NiuYawei Are the errors consistent? Do the same number happen every time? If not, we may have an issue. Otherwise, @sylviachanoiyee or @johannlombardi can force land it.
> > 
> > 
> > Yes, my observation is that pool create ULT will usually take more than 100 ms, and last CI run NLT also reported the same 5 new error messages.
> 
> It may be that we need to mask the time out from the error message, as we do with pointers currently but if the errors are consistent then it should be OK.

Sorry, I don't quite follow here, why do we need to mask the time out from the error message? I think printing the used time is useful for developer to know how long the ULT executed.

> 
> I assume this is the first step to identifying and resolving long-running threads? If so might it be more useful to dump the full stack in this case?

Yeah, dumping the full stack should be helpful, but I'm not sure how to do it. When scheduler detected a thread took too long in sched_watchdog_post(), the thread may have terminated and freed, if we save the full stack for every thread in sched_watchdog_prep(), I'm afraid that could be too heavy to impact performance. (and I didn't find ABT API for dumping stack to memory)

**Comment by liw** [comment]
I think we should land this useful functionality first, disabling the failed check that is preventing the landing.

**Comment by ashleypittman** [comment]
> > > > @NiuYawei Are the errors consistent? Do the same number happen every time? If not, we may have an issue. Otherwise, @sylviachanoiyee or @johannlombardi can force land it.
> > > 
> > > 
> > > Yes, my observation is that pool create ULT will usually take more than 100 ms, and last CI run NLT also reported the same 5 new error messages.
> > 
> > 
> > It may be that we need to mask the time out from the error message, as we do with pointers currently but if the errors are consistent then it should be OK.
> 
> Sorry, I don't quite follow here, why do we need to mask the time out from the error message? I think printing the used time is useful for developer to know how long the ULT executed.

When Jenkins reports warnings as a delta against a reference build it checks the text of the warning as well I believe, so NLT strips out things like pointer and uuids from the logs.  There's a danger with this PR that once landed future PRs would report five fixed warnings and five new ones.  I was proposing some ideas for avoiding this.

> > I assume this is the first step to identifying and resolving long-running threads? If so might it be more useful to dump the full stack in this case?
> 
> Yeah, dumping the full stack should be helpful, but I'm not sure how to do it. When scheduler detected a thread took too long in sched_watchdog_post(), the thread may have terminated and freed, if we save the full stack for every thread in sched_watchdog_prep(), I'm afraid that could be too heavy to impact performance. (and I didn't find ABT API for dumping stack to memory)

What is backtrace_symbols doing this?  I though that was giving the top entry on the stack, but if the interesting thread may have terminated and freed already then how does this help?

**Review by jolivier23 — DISMISSED** [review]

**Review by liw — DISMISSED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by liw — APPROVED** [review]

**Pull request merged**